### PR TITLE
provider/digitalocean: Add DO Cloud Firewalls (#15124)

### PR DESCRIPTION
**Pull request by keinohguchi**
This PR adds the DigitalOcean Cloud Firewalls support in digitalocean provider, as documented
the link below:

https://developers.digitalocean.com/documentation/v2/#firewalls

Here are a couple of links to the Cloud Firewalls blog posts:

https://www.digitalocean.com/products/cloud-firewalls/
https://blog.digitalocean.com/cloud-firewalls-secure-droplets-by-default/

and the actual commit to godo, DigitalOcean API binding in Go, for the CloudFirewalls support:

https://github.com/digitalocean/godo/commit/4fa9e9d999007b447089056a1c581b5594f0f851

And this is the output of the acceptance test:

```sh
air$ go test -v -run TestAccDigitalOceanFirewall
=== RUN   TestAccDigitalOceanFirewall_importBasic
=== RUN   TestAccDigitalOceanFirewall_importBasic/only_allow_inbound_SSH(TCP/22)
=== RUN   TestAccDigitalOceanFirewall_importBasic/only_allow_outbound_SSH(TCP/22)
=== RUN   TestAccDigitalOceanFirewall_importBasic/only_allow_inbound_SSH(TCP/22)_and_HTTP(TCP/80)
=== RUN   TestAccDigitalOceanFirewall_importBasic/only_allow_outbound_SSH(TCP/22)_and_DNS(UDP/53)
=== RUN   TestAccDigitalOceanFirewall_importBasic/allow_inbound_and_outbound_HTTPS(TCP/443),_inbound_SSH(TCP/22),_and_outbound_DNS(UDP/53)
--- PASS: TestAccDigitalOceanFirewall_importBasic (27.04s)
--- PASS: TestAccDigitalOceanFirewall_importBasic/only_allow_inbound_SSH(TCP/22) (4.97s)
--- PASS: TestAccDigitalOceanFirewall_importBasic/only_allow_outbound_SSH(TCP/22) (4.00s)
--- PASS: TestAccDigitalOceanFirewall_importBasic/only_allow_inbound_SSH(TCP/22)_and_HTTP(TCP/80) (8.94s)
--- PASS: TestAccDigitalOceanFirewall_importBasic/only_allow_outbound_SSH(TCP/22)_and_DNS(UDP/53) (5.18s)
--- PASS: TestAccDigitalOceanFirewall_importBasic/allow_inbound_and_outbound_HTTPS(TCP/443),_inbound_SSH(TCP/22),_and_outbound_DNS(UDP/53) (3.95s)
=== RUN   TestAccDigitalOceanFirewall_Basic
=== RUN   TestAccDigitalOceanFirewall_Basic/only_allow_inbound_SSH(TCP/22)
=== RUN   TestAccDigitalOceanFirewall_Basic/only_allow_outbound_SSH(TCP/22)
=== RUN   TestAccDigitalOceanFirewall_Basic/only_allow_inbound_SSH(TCP/22)_and_HTTP(TCP/80)
=== RUN   TestAccDigitalOceanFirewall_Basic/only_allow_outbound_SSH(TCP/22)_and_DNS(UDP/53)
=== RUN   TestAccDigitalOceanFirewall_Basic/allow_inbound_and_outbound_HTTPS(TCP/443),_inbound_SSH(TCP/22),_and_outbound_DNS(UDP/53)
--- PASS: TestAccDigitalOceanFirewall_Basic (24.62s)
--- PASS: TestAccDigitalOceanFirewall_Basic/only_allow_inbound_SSH(TCP/22) (4.35s)
--- PASS: TestAccDigitalOceanFirewall_Basic/only_allow_outbound_SSH(TCP/22) (5.53s)
--- PASS: TestAccDigitalOceanFirewall_Basic/only_allow_inbound_SSH(TCP/22)_and_HTTP(TCP/80) (4.60s)
--- PASS: TestAccDigitalOceanFirewall_Basic/only_allow_outbound_SSH(TCP/22)_and_DNS(UDP/53) (5.23s)
--- PASS: TestAccDigitalOceanFirewall_Basic/allow_inbound_and_outbound_HTTPS(TCP/443),_inbound_SSH(TCP/22),_and_outbound_DNS(UDP/53) (4.92s)
PASS ok github.com/hashicorp/terraform/builtin/providers/digitalocean 51.674s
air$
```

### CODE REVIEW

**Comment by stack72** [comment]
Hi @keinohguchi 

Thanks for the PR - please can you add some documentation to this PR as well?

Thanks

Paul

**Comment by keinohguchi** [comment]
Thank you, Paul, for the reply.  I've updated the documentation here with the link to the Cloud Firewalls APIv2.

Let me know if you need additional info, or different format.

Thank you for your support!

**Comment by keinohguchi** [comment]
ah, it's ```website/source/docs/providers/do/r```.  I'll add that and update the PR shortly.  Thank you for pointing that out, Paul!

**Comment by keinohguchi** [comment]
Hi Paul!

I've updated the PR with the documentation, as requested.

Here is the link to the rendered version, just in case:

https://github.com/keinohguchi/terraform/blob/master/website/source/docs/providers/aws/r/default_security_group.html.markdown#basic-example-usage-with-default-rules

Thank you for your review!

Cheers,

**Comment by iheanyi** [comment]
@keinohguchi – LGTM from the Firewalls implementation side of things 👍

File: builtin/providers/digitalocean/resource_digitalocean_firewall.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+		if resp != nil && resp.StatusCode == 404 {
+			log.Printf("[WARN] DigitalOcean Firewall (%s) not found", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving firewall: %s", err)
+	}
+
+	d.Set("name", firewall.Name)
```
Comment (stack72) [inline builtin/providers/digitalocean/resource_digitalocean_firewall.go]: For the import, we need to set *all* of the possible schema values that come back from the API

Comment (keinohguchi) [inline builtin/providers/digitalocean/resource_digitalocean_firewall.go]: Got you, Paul!  I'll add that and update the PR shortly!  Thank you for the thorough review, Paul!

**Comment by keinohguchi** [comment]
Got you, Paul, for the doc and test side for the import, too!  Thank you for the support!

File: builtin/providers/digitalocean/resource_digitalocean_firewall.go (hunk 1 above)
Comment (keinohguchi) [inline builtin/providers/digitalocean/resource_digitalocean_firewall.go]: Configurable fields, droplet_ids, tags, inboud_rules, and outbound_routes, are added.  It was a great exercise and now I do understand the criticality of this part of the code.  Thank you so much for your guidance, @stack72!

I'll add remaining status fields, more tests, documentation in regards to the firewall import capability.

I'll get back to you once everything in place.

Cheers,

File: website/source/layouts/digitalocean.erb [hunk 2]
```diff
@@ -33,6 +33,10 @@
           <a href="/docs/providers/do/r/droplet.html">digitalocean_droplet</a>
                     </li>
 
+                  <li<%= sidebar_current("docs-do-resource-firewall") %>>
+                    <a href="/docs/providers/do/r/firewall.html">digitalocean_firewall</a>
+                  </li>
+
                   <li<%= sidebar_current("docs-do-resource-floating-ip") %>>
```
Comment (keinohguchi) [inline website/source/layouts/digitalocean.erb]: digitalocean.erb side has been taken care of.  Thanks @stack72, for your guidance!  Really appreciated!

**Comment by keinohguchi** [comment]
Cool!  Import is up and running and the I put everything request above, code, test, doc, link, in the latest PR!

Please let me know if I miss anything, @stack72!

Cheers and thank you so much for your great support!

Here is the snippet of the simple import check in my local environment. (Unit test is there in the PR but I love to see it in action :)

```sh
air$ rm terraform.tfstate
air$ terraform show
No state.
air$ terraform import digitalocean_firewall.ImportFun b64d6efd-4352-4af5-98ad-09c58f86ed14
digitalocean_firewall.ImportFun: Importing from ID "b64d6efd-4352-4af5-98ad-09c58f86ed14"...
digitalocean_firewall.ImportFun: Import complete!
  Imported digitalocean_firewall (ID: b64d6efd-4352-4af5-98ad-09c58f86ed14)
digitalocean_firewall.ImportFun: Refreshing state... (ID: b64d6efd-4352-4af5-98ad-09c58f86ed14)

Import success! The resources imported are shown above. These are
now in your Terraform state. Import does not currently generate
configuration, so you must do this next. If you do not create configuration
for the above resources, then the next `terraform plan` will mark
them for destruction.
air$ terraform show
digitalocean_firewall.ImportFun:
  id = b64d6efd-4352-4af5-98ad-09c58f86ed14
  droplet_ids.# = 0
  inbound_rules.# = 3
  inbound_rules.0.port_range = 0
  inbound_rules.0.protocol = icmp
  inbound_rules.0.source_addresses.# = 2
  inbound_rules.0.source_addresses.0 = 0.0.0.0/0
  inbound_rules.0.source_addresses.1 = ::/0
  inbound_rules.0.source_droplet_ids.# = 0
  inbound_rules.0.source_load_balancer_uids.# = 0
  inbound_rules.0.source_tags.# = 0
  inbound_rules.1.port_range = 1-65535
  inbound_rules.1.protocol = tcp
  inbound_rules.1.source_addresses.# = 2
  inbound_rules.1.source_addresses.0 = 0.0.0.0/0
  inbound_rules.1.source_addresses.1 = ::/0
  inbound_rules.1.source_droplet_ids.# = 0
  inbound_rules.1.source_load_balancer_uids.# = 0
  inbound_rules.1.source_tags.# = 0
  inbound_rules.2.port_range = 1-65535
  inbound_rules.2.protocol = udp
  inbound_rules.2.source_addresses.# = 2
  inbound_rules.2.source_addresses.0 = 0.0.0.0/0
  inbound_rules.2.source_addresses.1 = ::/0
  inbound_rules.2.source_droplet_ids.# = 0
  inbound_rules.2.source_load_balancer_uids.# = 0
  inbound_rules.2.source_tags.# = 0
  name = server-999-allow-all
...
```

**Comment by keinohguchi** [comment]
Let me know if we need additional changes required, @stack72!  Thank you for your support!

**Comment by keinohguchi** [comment]
Hi @stack72!

Moved to the new provider, as below:

https://github.com/terraform-providers/terraform-provider-digitalocean/pull/1

Hense, closing this PR.

Thank you for your support, @stack72!

**Pull request closed without merging**